src/ucp/core/ucp_ep_vfs.c

```diff
@@ -0,0 +1,147 @@
+/**
+ * Copyright (C) 2021 NVIDIA CORPORATION & AFFILIATES. ALL RIGHTS RESERVED.
+ *
+ * See file LICENSE for terms.
+ */
+
+#ifdef HAVE_CONFIG_H
+#include "config.h"
+#endif
+
+#include "ucp_ep.inl"
+#include "ucp_ep_vfs.h"
+#include "ucp_vfs.h"
+
+#include <ucp/wireup/wireup_ep.h>
+#include <ucs/datastruct/string_buffer.h>
+#include <ucs/sys/sock.h>
+#include <ucs/vfs/base/vfs_cb.h>
+#include <ucs/vfs/base/vfs_obj.h>
+
+
+typedef struct {
+    const char *name;
+    uint64_t   field_mask;
+    size_t     offset;
+} ucp_ep_vfs_attr_t;
+
+
+static const ucp_ep_vfs_attr_t ucp_ep_vfs_attrs[] = {
+    {"local", UCP_EP_ATTR_FIELD_LOCAL_SOCKADDR, ucs_offsetof(ucp_ep_attr_t,
+                                                             local_sockaddr)},
+    {"remote", UCP_EP_ATTR_FIELD_REMOTE_SOCKADDR, ucs_offsetof(ucp_ep_attr_t,
+                                                               remote_sockaddr)}
+};
+
+static const char *ucp_err_handling_mode_names[] = {
+    [UCP_ERR_HANDLING_MODE_NONE] = "none",
+    [UCP_ERR_HANDLING_MODE_PEER] = "peer"
+};
+
+static void ucp_ep_vfs_read_peer_name(void *obj, ucs_string_buffer_t *strb,
+                                      void *arg_ptr, uint64_t arg_u64)
+{
+    ucp_ep_h ep = obj;
+
+    ucs_string_buffer_appendf(strb, "%s\n", ucp_ep_peer_name(ep));
+}
+
+static ucs_status_t ucp_ep_vfs_query_sockaddr(ucp_ep_h ep, ucp_ep_attr_t *attr,
+                                              uint64_t field_mask,
+                                              ucs_string_buffer_t *strb)
+{
+    ucs_status_t status;
+
+    attr->field_mask = field_mask;
+    status           = ucp_ep_query_sockaddr(ep, attr);
+    if (status != UCS_OK) {
+        ucs_string_buffer_appendf(strb, "<%s>", ucs_status_string(status));
+    }
+
+    return status;
+}
+
+static ucs_status_t ucp_ep_vfs_read_addr(ucp_ep_h ep, ucp_ep_vfs_attr_t *arg,
+                                         ucs_string_buffer_t *strb,
+                                         struct sockaddr *sockaddr)
+{
+    ucp_ep_attr_t attr;
+    ucs_status_t status;
+
+    status = ucp_ep_vfs_query_sockaddr(ep, &attr, arg->field_mask, strb);
+    if (status != UCS_OK) {
+        return status;
+    }
+
+    return ucs_sockaddr_copy(sockaddr, UCS_PTR_BYTE_OFFSET(&attr, arg->offset));
+}
+
+static void ucp_ep_vfs_read_addr_ip(void *obj, ucs_string_buffer_t *strb,
+                                    void *arg_ptr, uint64_t arg_u64)
+{
+    struct sockaddr sockaddr;
```

Comment:
1. have overflow here.
 either define struct sockaddr_storage ss;
or IMO better remove ucp_ep_vfs_read_addr helper function and thus get rid of the sock addr copy in line 76. define ucp_ep_attr_t attr; in ucp_ep_vfs_read_addr_ip/port.